### PR TITLE
Fix seeded stochastic completions in the server

### LINKED ISSUE #1245: mlx_lm.server: seed parameter silently ignored
## `mlx_lm.server`: `seed` parameter silently ignored

Chat-completions endpoint accepts `seed` and routes it through `mx.random.seed`, but output is the same across different seed values for the same prompt. mlx-lm 0.31.3, M-series, reproduced on Qwen2.5 and Llama-3.1 4 bit quants.

## Repro

Start server:

```bash
MODEL_PATH=/path/to/your/mlx/model
mlx_lm.server --model "$MODEL_PATH" --port 1234
```

Send three requests with different seeds (same shell, `$MODEL_PATH` carries over):

```bash
for s in 11111 22222 33333; do
  curl -sS http://127.0.0.1:1234/v1/chat/completions \
    -H 'Content-Type: application/json' \
    -d "{\"model\":\"$MODEL_PATH\",\"temperature\":1,\"max_tokens\":25,\"seed\":$s,\"stream\":false,\"messages\":[{\"role\":\"user\",\"content\":\"Tell me a one-sentence story about a cat.\"}]}" \
  | jq -r '.choices[0].message.content'
done
```
```
bash repro.sh
The tabby cat's reflection in the dark window was the only companion she hadn't driven away that evening.
The tabby cat's reflection in the dark window was the only companion she hadn't driven away that evening.
The tabby cat's reflection in the dark window was the only companion she hadn't driven away that evening.
```

All three responses come back byte-identical.

## Expected

Different seeds should follow different sampling trajectories which should provide different outputs (at temperature > 0).

## My thoughts why

`server.py:1191` reads `seed` from the body into `self.seed`. At `server.py:956-957` it's applied:

```python
if args.seed is not None:
    mx.random.seed(args.seed)
```

That call executes. Sampling lives at `sample_utils.py:267-282`:

```python
@partial(mx.compile, inputs=mx.random.state, outputs=mx.random.state)
def categorical_sampling(logits, temp):
    return mx.random.categorical(logits * (1 / temp))
```

~Hypothesis: `mx.compile` with `inputs=mx.random.state` captures the state object at decoration time, so subsequent `mx.random.seed(...)` calls don't propagate into what the compiled function reads. Effectively, sampling becomes deterministic against whatever state existed at module import.~
### Update

Initial suspicion was that the `mx.compile` decorator captured a stale `mx.random.state` reference. Not actually it: running `mx.random.seed(N) -> mx.random.categorical(...)` directly with three different seeds produces three different draws, both with and without the decorator. So it looks like the bug is somewhere downstream of the seed call but upstream of (or alongside) the actual sampling.
## Also tried

- Removed `seed` from the body, ran the same prompt twice. Identical output. State isn't naturally advancing across requests either.
- Same prompt at temperature 0.5 / 1.5 / 3.0. Outputs differ. Sampling runs, temperature is honored. Seed/state path is broken.
- Different prompts at fixed seed. Outputs differ. Not response-caching.

## Workaround

Vary something in the prompt per request (a hidden tag in the system message works). At least breaks the unintended determinism.

### PR DESCRIPTION
**Pull request by lyonsno**
## Summary

Fixes #1245.

The server accepts a `seed` parameter, but stochastic completions could still ignore per-request reseeding in a long-lived server process. The sampled path could keep using compiled random sampling across requests, and generated prompt-cache entries could replay a previous sampled assistant continuation for a later seeded request.

This keeps the default compiled sampler path for normal generation, but has the server use uncompiled random sampling only when a request is both seeded and stochastic. It also bypasses prompt-cache reuse for seeded stochastic requests to avoid reusing sampled assistant state across seed changes, while preserving the existing cache behavior for deterministic requests and unseeded stochastic requests.

## Changes

- Add a `use_compiled_sampling` option to `make_sampler()`, defaulting to the current compiled behavior.
- Add uncompiled random-sampling paths for categorical sampling and XTC so per-request `mx.random.seed()` is honored in the server's generation thread.
- Seed the server generation stream for requests that pass `seed`.
- Bypass prompt-cache fetch/insert for seeded stochastic requests only.
- Add regression coverage for threaded sampler seeding and server cache replay behavior.

## Verification

- `pytest tests/test_sample_utils.py::TestSampleUtils::test_uncompiled_sampler_respects_seed_in_thread tests/test_server.py::TestServer::test_handle_completions tests/test_server.py::TestServer::test_handle_completions_different_seeds_do_not_replay_cache tests/test_server.py::TestServer::test_prompt_cache_policy_only_bypasses_seeded_stochastic_requests -q`
- `pytest tests/test_sample_utils.py tests/test_server.py -q`
- `git diff --check`

### CODE REVIEW

**Comment by IsmaelMartinez** [comment]
Confirmed this fixes it. Tested on Apple Silicon, patched build is this branch installed editable, baseline is stock mlx_lm 0.31.3.

Same model (mlx-community/Qwen3-0.6B-4bit), same prompt, temperature 1.0, seeds 1/2/3 on /v1/chat/completions:

```
# stock 0.31.3, all three identical (the bug)
seed=1: A lighthouse stands tall and gleaming, its beacon casting long shadows...
seed=2: A lighthouse stands tall and gleaming, its beacon casting long shadows...
seed=3: A lighthouse stands tall and gleaming, its beacon casting long shadows...

# this branch, three different outputs
seed=1: A lighthouse stood on the edge of a stormy sea, its beacon humbly casting...
seed=2: The lighthouse stands as an eternal beacon of faith and purpose, its towers...
seed=3: A lighthouse stands tall in the darkness, its beam illuminating the pathways...
```

Same seed twice still returns identical output, so seeded stays reproducible while different seeds diverge, which is what you want. Tests pass on the branch too, the four you added plus the full test_sample_utils.py and test_server.py (30 passed). Would be good to see this land.

Thanks again!

**Comment by zcbenz** [comment]
This had been fixed by https://github.com/ml-explore/mlx/pull/3828 and I'm closing this in favor of https://github.com/ml-explore/mlx-lm/pull/1753.
